lib/commands/system-bars.js

```diff
@@ -0,0 +1,129 @@
+  result.y = parseFloat(frameMatch[2]);
+  const left = parseFloat(frameMatch[3]);
+  const top = parseFloat(frameMatch[4]);
+  result.width = left - result.x;
+  result.height = top - result.y;
+  const visibilityMatch = SURFACE_PATTERN.exec(propLines);
+  if (!visibilityMatch) {
+    log.debug(propLines);
+    throw new Error(`Cannot parse the visibility value from '${name}' window properties`);
+  }
+  result.visible = visibilityMatch[1] === 'true';
+  return result;
+}
+
+/**
+ * Extracts status and navigation bar information from the window manager output.
+ *
+ * @param {Array<string>} lines Output from dumpsys command.
+ * Check the corresponding unit tests for more details on the input format.
+ * @return {Object} An object containing two items where keys are statusBar and navigationBar,
+ * and values are corresponding WindowProperties objects
+ * @throws {Error} If no window properties could be parsed
+ */
+function parseWindows (lines) {
+  const windows = {};
+  let currentWindowName = null;
+  let windowNameRowIndent = null;
+  for (const line of lines.split('\n').map(_.trimEnd)) {
+    const currentIndent = line.length - _.trimStart(line).length;
+    if (_.isNil(windowNameRowIndent) || currentIndent <= windowNameRowIndent) {
+      const match = WINDOW_TITLE_PATTERN.exec(line);
+      if (!match) {
+        currentWindowName = null;
+        continue;
+      }
+      currentWindowName = match[1];
+      if (_.isNil(windowNameRowIndent)) {
+        windowNameRowIndent = currentIndent;
+      }
+      continue;
+    }
+    if (!currentWindowName || currentIndent <= windowNameRowIndent) {
+      continue;
+    }
+
+    if (!_.isArray(windows[currentWindowName])) {
+      windows[currentWindowName] = [];
+    }
+    windows[currentWindowName].push(line);
+  }
+  if (_.isEmpty(windows)) {
+    log.debug(lines.join('\n'));
+    throw new Error('Cannot parse any window information from the dumpsys output');
+  }
+
+  const result = {statusBar: null, navigationBar: null};
+  for (const [name, props] of _.toPairs(windows)) {
+    if (name.startsWith(STATUS_BAR_WINDOW_NAME_PREFIX)) {
+      result.statusBar = parseWindowProperties(name, props);
+    } else if (name.startsWith(NAVIGATION_BAR_WINDOW_NAME_PREFIX)) {
+      result.navigationBar = parseWindowProperties(name, props);
+    }
+  }
+  const unmatchedWindows = [
+    ['statusBar', STATUS_BAR_WINDOW_NAME_PREFIX],
+    ['navigationBar', NAVIGATION_BAR_WINDOW_NAME_PREFIX]
+  ].filter(([name]) => _.isNil(result[name]));
+  for (const [window, namePrefix] of unmatchedWindows) {
+    log.info(`No windows have been found whose title matches to ` +
+      `'${namePrefix}'. Assuming it is invisible. ` +
+      `Only the following windows are available: ${_.keys(windows)}`);
+    result[window] = _.cloneDeep(DEFAULT_WINDOW_PROPERTIES);
+  }
+  return result;
+}
+
+commands.getSystemBars = async function getSystemBars () {
+  const out = await this.adb.shell(['dumpsys', 'window', 'windows']);
```

Comment:
need try/catch? e.g. when the device is disconnected at this time

Reply:
added